CMakeLists.txt

```diff
@@ -0,0 +1,101 @@
+set(FINUFFT_FFTW_SUFFIX "OpenMP" CACHE STRING "Suffix for FFTW libraries (e.g. OpenMP, Threads etc.)")
+set(FINUFFT_BUILD_EXAMPLES "OFF" CACHE BOOL "Whether to build the FINUFFT examples")
+set(FINUFFT_BUILD_TESTS "OFF" CACHE BOOL "Whether to build the FINUFFT tests")
+set(FINUFFT_BUILD_FORTRAN "OFF" CACHE BOOL "Whether to build the FINUFFT Fortran examples")
+set(FINUFFT_BUILD_MATLAB "OFF" CACHE BOOL "Whether to build the FINUFFT Matlab interface")
+set(FINUFFT_ENABLE_SANITIZERS "ON" CACHE BOOL "Whether to enable sanitizers, only effective for Debug configuration.")
+
+if(FINUFFT_BUILD_MATLAB)
+    # When building for matlab, we will fetch the OpenMP library used by matlab
+    # instead of system default for compatibility.
+    find_package(Matlab REQUIRED)
+    find_library(OpenMP_iomp5_LIBRARY NAMES iomp5 HINTS ${Matlab_ROOT_DIR}/sys/os/ PATH_SUFFIXES glnxa64 maci64)
+    find_library(OpenMP_pthreads_LIBRARY NAMES pthread CMAKE_FIND_ROOT_PATH_BOTH)
+endif()
+
+# Find required packages
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
+find_package(FFTW REQUIRED)
+find_package(OpenMP REQUIRED)
+
+set(FINUFFT_FFTW_LIBRARIES "FFTW::Float" "FFTW::Double" "FFTW::Float${FINUFFT_FFTW_SUFFIX}" "FFTW::Double${FINUFFT_FFTW_SUFFIX}")
+
+# This set of sources is compiled twice, once in single precision and once in double precision
+# The single precision compilation is done with -DSINGLE
+set(FINUFFT_PRECISION_DEPENDENT_SOURCES src/finufft.cpp src/simpleinterfaces.cpp src/spreadinterp.cpp src/utils.cpp fortran/finufftfort.cpp)
+
+# Utility function to enable ASAN on debug builds
+set(FINUFFT_ASAN_SUPPORTED_COMPILERS AppleClang Clang GNU)
+function(enable_asan target)
+    if(NOT FINUFFT_ENABLE_SANITIZERS)
+        return()
+    endif()
+
+    if(CMAKE_CXX_COMPILER_ID IN_LIST FINUFFT_ASAN_SUPPORTED_COMPILERS)
+        # Enable only on clang / gcc compilers.
+        target_compile_options(${target} PUBLIC $<$<CONFIG:DEBUG>:-fsanitize=address>)
+        target_link_options(${target} PUBLIC $<$<CONFIG:DEBUG>:-fsanitize=address>)
+    endif()
+endfunction()
+
+# Utility function to set finufft compilation options.
+function(set_finufft_options target)
+    set_property(TARGET ${target} PROPERTY POSITION_INDEPENDENT_CODE ON)
+    set_property(TARGET ${target} PROPERTY CMAKE_CXX_STANDARD 14)
+    enable_asan(${target})
+    target_compile_options(${target} PRIVATE SHELL:$<$<CONFIG:Release,RelWithDebInfo>:${FINUFFT_ARCH_FLAGS}>)
+    target_include_directories(${target} PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}/include")
+    target_link_libraries(${target} PUBLIC OpenMP::OpenMP_CXX)
+endfunction()
+
+# Main finufft libraries
+add_library(finufft_f32 OBJECT ${FINUFFT_PRECISION_DEPENDENT_SOURCES})
+target_compile_definitions(finufft_f32 PRIVATE SINGLE)
+set_finufft_options(finufft_f32)
+target_link_libraries(finufft_f32 PUBLIC ${FINUFFT_FFTW_LIBRARIES})
```

Comment:
this should be private. The user will not use fftw or openmp via finufft, if they want to, they can link against them separately

Reply:
Thanks for spotting this! I will look into visibility here. `target_link_libraries` is required as it also sets include flags (not just link flags), which are required for object libraries nonetheless.

Reply:
Yes, but 
```cmake
find_package(FFTW REQUIRED)
```
does ``` FFTW_INCLUDE_DIRS ... fftw include directory paths```
So you can just:
```cmake  
target_include_directories(${target} PRIVATE "${FFTW_INCLUDE_DIRS }")
```

Reply:
Of course, but `target_link_library` expresses a semantic dependency, which is much easier. The dependency also needs to be public at the moment as `finufft` headers include `fftw` headers (at least for now), so include paths must be transitively set.

Reply:
Yes, the problem is that you should linke fftw here: 
```cmake
target_link_libraries(finufft_static PUBLIC finufft_f32 finufft_f64 m ${FINUFFT_FFTW_LIBRARIES})
```
not inside the object library.